elide-core/src/main/java/com/yahoo/elide/security/permissions/ExpressionBuilder.java

```diff
@@ -0,0 +1,328 @@
+
+            Expression operationExpression = new AnyFieldExpression(
+                    buildExpression(entityChecks, opCheckFn, fieldClassJoiner),
+                    buildExpression(fieldChecks, opCheckFn, fieldJoiner)
+            );
+
+            entityChecks = arrayToQueue(extracted.getEntityChecks());
+            fieldChecks = arrayToQueue(extracted.getFieldChecks());
+
+            Expression commitExpression = new AnyFieldExpression(
+                    buildExpression(entityChecks, comCheckFn, fieldClassJoiner),
+                    buildExpression(fieldChecks, comCheckFn, fieldJoiner)
+            );
+
+            opExp = new OrExpression(opExp, operationExpression);
+            comExp = new OrExpression(comExp, commitExpression);
+        }
+
+        return new Expressions(opExp, comExp);
+    }
+
+    /**
+     * Build a specific expression for a check.
+     *
+     * @param checks Checks to build expression for
+     * @param expressionBuilder Builder method to convert check to an expression
+     * @param expressionJoiner Method to join checks
+     * @return Expression
+     */
+    private Expression buildExpression(final Queue<Class<? extends Check>> checks,
+                                       final Function<Check, Expression> expressionBuilder,
+                                       final BiFunction<Expression, Expression, Expression> expressionJoiner) {
+        if (checks.size() == 0) {
+            return null;
+        }
+
+        try {
+            Check instance = checks.poll().newInstance();
+            Expression expression = expressionBuilder.apply(instance);
+            return expressionJoiner.apply(expression, buildExpression(checks, expressionBuilder, expressionJoiner));
+        } catch (InstantiationException | IllegalAccessException e) {
+            log.error("Could not access check. Exception: {}", e);
+            throw new InvalidSyntaxException("Could not instantiate specified check.");
+        }
+    }
+
+    /**
+     * Retrieve the proper joiner for a specified check mode.
+     *
+     * @param mode Mode
+     * @return Joiner
+     */
+    private static BiFunction<Expression, Expression, Expression> getJoiner(ExtractedChecks.CheckMode mode) {
+        return (mode == ALL) ? ALL_JOINER : ANY_JOINER;
+    }
+
+    /**
+     * Convert an array to queue.
+     *
+     * @param array Array to convert
+     * @param <T> Type parameter
+     * @return Queue representing array. Empty queue if array is null.
+     */
+    private static <T> Queue<T> arrayToQueue(final T[] array) {
+        final Queue<T> queue = new LinkedBlockingQueue<>();
+        if (array == null) {
+            return queue;
+        }
+        for (T value : array) {
+            queue.add(value);
+        }
+        return queue;
+    }
+
+    /**
+     * Structure containing built expressions.
+     */
+    @AllArgsConstructor
+    public static class Expressions {
```

Comment:
Not sure having the return types bundled here helps readability.  Maybe it would be better to for the builder to be passed in the provider or factory method that creates the underlying leaf expression objects - and then let the calling functions call this twice - once for operation and once for commit.